### DISCUSSION TITLE
I want to sort SQL pagequeries naturally

### DISCUSSION THREAD

**Discussion by vaniwiki** [combostrap/combo#100] [Q&A]
I am making a view that shows me a card for each chapter within that book. These chapters are called 1:start, 2:start, 3:start and so on, up to twenty, thirty chapters or more. Which means they will display 1:start, 10:start, 11:start... I would like to be able to sort them naturally 1, 2, 3 before 10, 11, 12. Just wondering if anyone knows.

```
<iterator>
    <data>select where path like '%books:bg:%' and path like '%start%' order by path asc</data>
    <card-columns>
        <fragment>
            <card clickable>

            === $label ===

            ${description | head(1024,'...')} [[$path?preview=false| ]]
            </card>
        </fragment>
    </card-columns>
</iterator>
```

**Comment by gerardnico** [comment]
It should work. The engine is SQLLite therefore this is the sorting of Sqlite.
See doc: https://sqlite.org/datatype3.html#sorting_grouping_and_compound_selects

**Reply by vaniwiki** [reply]
Hmm, not really. I did a search on natural sorting in SQLite, and it seems natural sorting is not available by default. One has to make one's own rulesets or something... is what I garnered from here: https://sqlite.org/forum/info/d5cf6c6317dd7e7f

**Reply by gerardnico** [reply]
Man what do you mean with natural. If you got only numbers as characters, they should be in order.

What can happen is that the numbers are before the characters and vice versa.
It all depends on the character set and the collation (ie comparison)
If you got only numbers in the first place, you have no problem.

Show me your problem for now you just show the code and it's really not funny when someone tries to help you.

Anyway, I can't change that easily therefore, you can may be solve that with a filter for the number and a filter for the characters on 2 blocks.

**Reply by gerardnico** [reply]
ie I assume that you get this order

``` 
1:start
10:start
2:start
```
and you want
``` 
1:start
2:start
10:start
```
That's because `:` is before `0` in the character set.

This should work:
``` 
01:start
02:start
10:start
```

**Reply by vaniwiki** [reply]
It will sort like this:
```
1
11
111
12
13
2
3
4
```

Don't think tha path colons have anything to do with it, necessarily. It outputs the character 111 before 2. But naturally number 2 comes before 111 in any person's mind. That is what I want. 

Your solution would indeed work, if such was my folder/page structure. Alas, such it is not.

But in 2024, I think natural sorting should be trivial, especially for complex query languages. Like you pointed out, in my case it would be possible to do programmatically by just adding appropriate zeroes in front so that all pagenames become equal length. Something like that. But best if this could be done in the query, and not change folder structure.

I am open to alternate solutions. I tried setting date_published to 1960-01-01, 1960-01-02 and so on, to "override." But I could not figure out how to get those records and display them... Still new to this, learning the ropes. One step at a time.

**Reply by gerardnico** [reply]
As I said, natural sorting is just a name taken from the fact that you see them as numbers but they are not, they are characters

You must see them like that
```
a
aa
aaa
ab
ac
b
c
d
```
And that's not what you will get, you will get
```
a
b
c
d
aa
ab
ac
aaa
```
More natural?

There is no natural with text.
See: https://datacadamia.com/data/type/text/order

**Reply by gerardnico** [reply]
The only way would be to have a numeric sort field but Combo does not have one.

**Reply by vaniwiki** [reply]
Not even the date_published, and work in incremental date values?

**Reply by gerardnico** [reply]
I don't understand what you wrote and what you want to do.

giving examples with your writing helps your message to be better brought over.

I have other things to do today. Therefore pardon my future silence. Wish you the best.

**Reply by vaniwiki** [reply]
Sorry. I don't mean to take so much of your time. This information I should be able to glean from other places...

On [your page here](https://combostrap.com/templating-markup/how-to-define-your-data-set-with-sql-55fkwj0m#attribute) it seems you are saying that the attributes, or variables(?) on [this page](https://combostrap.com/templating-markup/the-variables-that-can-be-used-anywhere-qfr74w67#variable_list) can be used to sort by. I assumed that I could use the date_published attribute and sort by that. Gives me plenty of numbers to work with, days, months, even years.

Speech is silver, and silence is golden. Wish you a blessed day. :)

**Reply by vaniwiki** [reply]
It actually worked to sort with the date_published attribute. The fault lay in my query somewhere. Now it reads:

`<data path=":books:lob:">select from descendants order by date_published asc</data>`

Which works. It is cumbersome to do all pages like that, but maybe if I set a secondary sort option, I can get away with having to set date_published only on the first nine pages. That is, if total pages are under 100. Else, the first 99. Let me try...

As long as I can find a workaround, I'm not gonna complain. Just sharing. :)
